exercises/crypto-square/example.nim

```diff
@@ -0,0 +1,21 @@
+import math, strutils
+
+func encrypt*(s: string): string =
+  var normalized = newStringOfCap(s.len)
+  for c in s.toLowerAscii():
+    if c in {'a'..'z'} or c in {'0'..'9'}:
+      normalized &= c
+
+  let r = normalized.len.float.sqrt.ceil.int
+  let c = (normalized.len.float / r.float).round.int
+  normalized = normalized.alignLeft(r*c, ' ')
+
+  var ct = newSeq[string]()
+  for s in 0 ..< r:
+    var word = ""
+    for i in countup(s, normalized.high, r):
+      word &= normalized[i]
+
+    ct &= word
+
+  result = ct.join(" ")
```

Comment:
Could you just return `ct.join(" ")`? I'm not seeing `result` used anywhere else in the code, unless I'm missing it.

Reply:
> Could you just return `ct.join(" ")`?

### Short answer
Yes, `ct.join(" ")` also works, but I think `result = ct.join(" ")` is fine here because:
- The Nim style guide doesn't explicitly mention `result =`
- `result =` is maybe the de-facto standard anyway
- The `meetup` example solution sets a precedent
- It raises similar questions of style such as `.toStr` vs `.toStr()`

However, I don't mind changing it if you want. Thanks for noticing.


### Long answer
The [Nim style guide](https://nim-lang.org/docs/nep1.html#introduction-coding-conventions) prefers `result` over `return` unless control-flow is required, but doesn't explicitly say whether to use `result =` when it is optional.

In my own code I tend to use `result =` where it improves clarity (e.g. in longer or potentially confusing functions), and omit it where it adds noise (e.g. in one-liners or `case` statements).


#### Style in `exercism/nim`
It is indeed rare in this repo to use `result =` when it is optional. The only comparable occurrence is in [meetup/example.nim](https://github.com/exercism/nim/blob/3cbf374e451634c53124c29403818f4861ceb881/exercises/meetup/example.nim#L14), where I think it does improve clarity.

One example that omits `result =` is [sublist/example.nim](https://github.com/exercism/nim/blob/3cbf374e451634c53124c29403818f4861ceb881/exercises/sublist/example.nim#L21), where I write:
```Nim
func sublist*(a, b: openArray[int]): Comparison =
  if a == b:
    Equal
  elif a.contains(b):
    Superlist
  elif b.contains(a):
    Sublist
  else:
    Unequal
```

Rather than:
```Nim
func sublist*(a, b: openArray[int]): Comparison =
  if a == b:
    result = Equal
  elif a.contains(b):
    result = Superlist
  elif b.contains(a):
    result = Sublist
  else:
    result = Unequal
```

Or an `if expression` with its extra indentation layer:
```Nim
func sublist*(a, b: openArray[int]): Comparison =
  result =
    if a == b:
      Equal
    elif a.contains(b):
      Superlist
    elif b.contains(a):
      Sublist
    else:
      Unequal
```

Another example of omitting `result =` is [resistor-color/example.nim](https://github.com/exercism/nim/blob/3cbf374e451634c53124c29403818f4861ceb881/exercises/resistor-color/example.nim#L5)
```Nim
func colorCode*(color: ResistorColor): int =
  color.ord
```


#### Style in `nim-lang/Nim`
I was curious how strongly the Nim repo prefers one style, so I had a look. It seems like `result =` is de-facto strongly preferred.

I didn't see many counterexamples, but there are some in e.g. `strutils` and `times`:

The following use `result =`
- [compiler/ast.nim](https://github.com/nim-lang/Nim/blob/041d15392aaf732665abab290f0cf5993d909efc/compiler/ast.nim#L1028)
- [compiler/bitsets.nim](https://github.com/nim-lang/Nim/blob/041d15392aaf732665abab290f0cf5993d909efc/compiler/bitsets.nim#L34)
- [compiler/incremental.nim](https://github.com/nim-lang/Nim/blob/041d15392aaf732665abab290f0cf5993d909efc/compiler/incremental.nim#L69)
- [compiler/parampatterns.nim](https://github.com/nim-lang/Nim/blob/041d15392aaf732665abab290f0cf5993d909efc/compiler/parampatterns.nim#L51)
- [lib/system.nim](https://github.com/nim-lang/Nim/blob/041d15392aaf732665abab290f0cf5993d909efc/lib/system.nim#L1574)
- [lib/system/io.nim](https://github.com/nim-lang/Nim/blob/041d15392aaf732665abab290f0cf5993d909efc/lib/system/io.nim#L149)
- [lib/system/nimscript.nim](https://github.com/nim-lang/Nim/blob/041d15392aaf732665abab290f0cf5993d909efc/lib/system/nimscript.nim#L158)
- [lib/system/osalloc.nim](https://github.com/nim-lang/Nim/blob/041d15392aaf732665abab290f0cf5993d909efc/lib/system/osalloc.nim#L260)
- [lib/system/sets.nim](https://github.com/nim-lang/Nim/blob/041d15392aaf732665abab290f0cf5993d909efc/lib/system/sets.nim#L17)
- [lib/system/threads.nim](https://github.com/nim-lang/Nim/blob/041d15392aaf732665abab290f0cf5993d909efc/lib/system/threads.nim#L189)
- [lib/pure/algorithm.nim](https://github.com/nim-lang/Nim/blob/041d15392aaf732665abab290f0cf5993d909efc/lib/pure/algorithm.nim#L51)
- [lib/pure/base64.nim](https://github.com/nim-lang/Nim/blob/041d15392aaf732665abab290f0cf5993d909efc/lib/pure/base64.nim#L144)
- [lib/pure/httpclient.nim](https://github.com/nim-lang/Nim/blob/041d15392aaf732665abab290f0cf5993d909efc/lib/pure/httpclient.nim#L176)
- [lib/pure/math.nim](https://github.com/nim-lang/Nim/blob/041d15392aaf732665abab290f0cf5993d909efc/lib/pure/math.nim#L687)
- [lib/pure/nativesockets.nim](https://github.com/nim-lang/Nim/blob/041d15392aaf732665abab290f0cf5993d909efc/lib/pure/nativesockets.nim#L172)
- [lib/pure/parsecfg.nim](https://github.com/nim-lang/Nim/blob/041d15392aaf732665abab290f0cf5993d909efc/lib/pure/parsecfg.nim#L180)
- [lib/pure/punycode.nim](https://github.com/nim-lang/Nim/blob/041d15392aaf732665abab290f0cf5993d909efc/lib/pure/punycode.nim#L38)
- [lib/pure/random.nim](https://github.com/nim-lang/Nim/blob/041d15392aaf732665abab290f0cf5993d909efc/lib/pure/random.nim#L338)
- [lib/pure/strutils.nim](https://github.com/nim-lang/Nim/blob/041d15392aaf732665abab290f0cf5993d909efc/lib/pure/strutils.nim#L207)
- [lib/pure/strutils.nim](https://github.com/nim-lang/Nim/blob/041d15392aaf732665abab290f0cf5993d909efc/lib/pure/strutils.nim#L1193)
- [lib/pure/strutils.nim](https://github.com/nim-lang/Nim/blob/041d15392aaf732665abab290f0cf5993d909efc/lib/pure/strutils.nim#L1438)
- [lib/pure/times.nim](https://github.com/nim-lang/Nim/blob/041d15392aaf732665abab290f0cf5993d909efc/lib/pure/times.nim#L1258)
- [lib/pure/collections/critbits.nim](https://github.com/nim-lang/Nim/blob/041d15392aaf732665abab290f0cf5993d909efc/lib/pure/collections/critbits.nim#L36)
- [lib/pure/collections/deques.nim](https://github.com/nim-lang/Nim/blob/041d15392aaf732665abab290f0cf5993d909efc/lib/pure/collections/deques.nim#L80)
- [lib/pure/collections/intsets.nim](https://github.com/nim-lang/Nim/blob/041d15392aaf732665abab290f0cf5993d909efc/lib/pure/collections/intsets.nim#L53)
- [lib/pure/collections/lists.nim](https://github.com/nim-lang/Nim/blob/041d15392aaf732665abab290f0cf5993d909efc/lib/pure/collections/lists.nim#L334)
- [lib/pure/collections/sets.nim](https://github.com/nim-lang/Nim/blob/041d15392aaf732665abab290f0cf5993d909efc/lib/pure/collections/sets.nim#L640)
- [lib/pure/collections/tables.nim](https://github.com/nim-lang/Nim/blob/041d15392aaf732665abab290f0cf5993d909efc/lib/pure/collections/tables.nim#L502)
- [lib/std/editdistance.nim](https://github.com/nim-lang/Nim/blob/041d15392aaf732665abab290f0cf5993d909efc/lib/std/editdistance.nim#L264)

And these omit `result =`
- [lib/pure/asyncfile.nim](https://github.com/nim-lang/Nim/blob/041d15392aaf732665abab290f0cf5993d909efc/lib/pure/asyncfile.nim#L40) (used in this proc, but omitted in the simpler one below it)
- [lib/pure/strutils.nim](https://github.com/nim-lang/Nim/blob/041d15392aaf732665abab290f0cf5993d909efc/lib/pure/strutils.nim#L427)
- [lib/pure/times.nim](https://github.com/nim-lang/Nim/blob/041d15392aaf732665abab290f0cf5993d909efc/lib/pure/times.nim#L507)


#### Comment
Nim's flexibility raises other style questions that I think we can ignore, at least for now.

For example, in my own code I prefer `.toStr` to `.toStr()`, but we could argue that `.toStr()` is better for this repo because `.toStr` isn't syntax-highlighted in popular editors or on GitHub.

Reply:
All this sounds good, thanks for taking the time to look into this! No change needed. 😄